--- a/src/_portage	2015-03-07 19:13:55.000000000 +0900
+++ b/src/_portage	2015-03-07 19:13:48.000000000 +0900
@@ -92,7 +92,7 @@
     "($noask_opts)"{-A,--alert}"[Add a terminal bell to all interactive prompts]"
     "--ignore-default-opts[Ignore EMERGE_DEFAULT_OPTS]"
     "--moo[Have you mooed today?]"
-    "(--jobs)--jobs[Number of packages to build simultaneously]:jobs:({1.."${#${$(</proc/cpuinfo)/^processor}}"})"
+    "(--jobs)--jobs[Number of packages to build simultaneously]:jobs:({1..n})"
   )
 
   install_args=(
