diff -Naur vlc.orig/vlc-1.1.11/libs/unzip/ioapi.h vlc.new/vlc-1.1.11/libs/unzip/ioapi.h
--- vlc.orig/vlc-1.1.11/libs/unzip/ioapi.h	2010-04-13 02:22:26.000000000 +0200
+++ vlc.new/vlc-1.1.11/libs/unzip/ioapi.h	2011-09-18 00:25:13.000000000 +0200
@@ -9,6 +9,9 @@
 #ifndef _ZLIBIOAPI_H
 #define _ZLIBIOAPI_H
 
+#ifndef OF
+#define OF _Z_OF
+#endif
 
 #define ZLIB_FILEFUNC_SEEK_CUR (1)
 #define ZLIB_FILEFUNC_SEEK_END (2)
