--- jd-2.8.0-110203.orig/src/global.h	2011-03-04 22:48:50.018261890 +0900
+++ jd-2.8.0-110203/src/global.h	2011-03-04 22:49:12.098261890 +0900
@@ -219,12 +219,12 @@
 
 enum
 {
-    AUTORELOAD_MINSEC = 10, // オートリロードの最小秒数
+    AUTORELOAD_MINSEC = 1, // オートリロードの最小秒数
     MIN_LIVE_RELOAD_SEC = AUTORELOAD_MINSEC, // 実況時の最小リロード間隔
 
-    WAITLOADIMG_SEC = 6, // 画像のロード待ち間隔
+    WAITLOADIMG_SEC = 0, // 画像のロード待ち間隔
 
-    CHECKUPDATE_MINSEC = 300 // 更新チェックの最小秒数
+    CHECKUPDATE_MINSEC = 1 // 更新チェックの最小秒数
 };
 
 
