diff --git a/src/emuMB.c b/src/emuMB.c
index b25eac8..085c135 100644
--- a/src/emuMB.c
+++ b/src/emuMB.c
@@ -191,7 +191,7 @@ EvdevMBEmuTimer(InputInfoPtr pInfo)
 
     pEvdev->emulateMB.pending = FALSE;
     if ((id = stateTab[pEvdev->emulateMB.state][4][0]) != 0) {
-        EvdevPostButtonEvent(pInfo, abs(id),
+        EvdevPostButtonEvent(pInfo, pEvdev->btnmap[abs(id)],
                              (id >= 0) ? BUTTON_PRESS : BUTTON_RELEASE);
         pEvdev->emulateMB.state =
             stateTab[pEvdev->emulateMB.state][4][2];
@@ -237,12 +237,12 @@ EvdevMBEmuFilterEvent(InputInfoPtr pInfo, int button, BOOL press)
 
     if ((id = stateTab[pEvdev->emulateMB.state][*btstate][0]) != 0)
     {
-        EvdevQueueButtonEvent(pInfo, abs(id), (id >= 0));
+        EvdevQueueButtonEvent(pInfo, pEvdev->btnmap[abs(id)], (id >= 0));
         ret = TRUE;
     }
     if ((id = stateTab[pEvdev->emulateMB.state][*btstate][1]) != 0)
     {
-        EvdevQueueButtonEvent(pInfo, abs(id), (id >= 0));
+        EvdevQueueButtonEvent(pInfo, pEvdev->btnmap[abs(id)], (id >= 0));
         ret = TRUE;
     }
 
