diff --git a/Makefile.in b/Makefile.in
index 903b016..a1e39ae 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -32,7 +32,8 @@ PACKAGE   = gauche-vim-xire
 ARCHFILES =
 SCMFILES  = $(srcdir)/vim/xire.scm \
             $(srcdir)/vim/xire/builtin.scm \
-            $(srcdir)/vim/xire/compiler.scm
+            $(srcdir)/vim/xire/compiler.scm \
+            $(srcdir)/vim/xire/ivs.scm
 HEADERS   = 
 
 TARGET    = $(ARCHFILES)
