--- a/src/screen.C	2011-07-02 20:23:49.621010111 +0900
+++ b/src/screen.C	2011-07-02 20:23:17.207010088 +0900
@@ -2420,8 +2420,8 @@
                 XSetForeground (dpy, gc, pix_colors[fore]);
 
               XDrawLine (dpy, vt, gc,
-                         xpixel, ypixel + font->ascent + 1,
-                         xpixel + Width2Pixel (count) - 1, ypixel + font->ascent + 1);
+                         xpixel, ypixel + Height2Pixel (1) - 1,
+                         xpixel + Width2Pixel (count) - 1, ypixel + Height2Pixel (1) - 1);
             }
         }                     /* for (col....) */
     }                         /* for (row....) */
